File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        i=0
        n=len(s)
        intmax=2**31-1
        inmin=-2**31
        while i<n and s[i]==' ':
            i+=1
        sign=1
        if i<n and s[i] in ["+","-"]:
            if s[i]=="-":
                sign=-1
            i+=1
        res=0
        while i<n and s[i].isdigit():
            digit=int(s[i])
            res=res*10+digit
            i+=1
        res*=sign
        if res<inmin:
            return inmin
        if res>intmax:
            return intmax
        return res

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.go
func myAtoi(s string) int {
	const INT_MAX = 1<<31 - 1
	const INT_MIN = -1 << 31

	n := len(s)
	i := 0

	// 1. 跳过前导空格
	for i < n && s[i] == ' ' {
		i++
	}

	// 2. 处理符号
	sign := 1
	if i < n && (s[i] == '+' || s[i] == '-') {
		if s[i] == '-' {
			sign = -1
		}
		i++
	}

	// 3. 读取连续数字
	res := 0

	for i < n && s[i] >= '0' && s[i] <= '9' {
		digit := int(s[i] - '0')

		// 4. 提前判断溢出
		if sign == 1 {
			if res > INT_MAX/10 || (res == INT_MAX/10 && digit > 7) {
				return INT_MAX
			}
		} else {
			if -res < INT_MIN/10 || (-res == INT_MIN/10 && digit > 8) {
				return INT_MIN
			}
		}

		res = res*10 + digit
		i++
	}

	return sign * res
}